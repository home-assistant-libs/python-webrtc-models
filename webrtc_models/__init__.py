"""WebRTC models."""

from dataclasses import dataclass, field
from typing import Any

from mashumaro import field_options
from mashumaro.config import BaseConfig
from mashumaro.mixins.orjson import DataClassORJSONMixin

__all__ = [
    "RTCConfiguration",
    "RTCIceCandidate",
    "RTCIceServer",
]


class _RTCBaseModel(DataClassORJSONMixin):
    """Base class for RTC models."""

    class Config(BaseConfig):
        """Mashumaro config."""

        # Serialize to spec conform names and omit default values
        omit_default = True
        serialize_by_alias = True


@dataclass
class RTCIceServer(_RTCBaseModel):
    """RTC Ice Server.

    See https://www.w3.org/TR/webrtc/#rtciceserver-dictionary
    """

    urls: str | list[str]
    username: str | None = None
    credential: str | None = None


@dataclass
class RTCConfiguration(_RTCBaseModel):
    """RTC Configuration.

    See https://www.w3.org/TR/webrtc/#rtcconfiguration-dictionary
    """

    ice_servers: list[RTCIceServer] = field(
        metadata=field_options(alias="iceServers"), default_factory=list
    )


@dataclass(frozen=True)
class RTCIceCandidateInit(_RTCBaseModel):
    """RTC Ice Candidate Init.

    See https://www.w3.org/TR/webrtc/#rtcicecandidate-interface
    """

    candidate: str
    sdp_m_line_index: int | None = field(
        metadata=field_options(alias="sdpMLineIndex"), default=None, kw_only=True
    )
    sdp_mid: str | None = field(
        metadata=field_options(alias="sdpMid"), default=None, kw_only=True
    )
    user_fragment: str | None = field(
        metadata=field_options(alias="userFragment"), default=None, kw_only=True
    )


@dataclass(frozen=True)
class RTCIceCandidate(_RTCBaseModel):
    """RTC Ice Candidate.

    See https://www.w3.org/TR/webrtc/#rtcicecandidate-interface
    """

    class Config(BaseConfig):
        """Mashumaro config."""

        # Serialize to spec conform names and omit default values
        omit_default = False
        serialize_by_alias = True

    rtc_ice_candidate_init: RTCIceCandidateInit | None = field(
        default=None, metadata=field_options(serialize="omit")
    )

    candidate: str = field(init=False)
    sdp_m_line_index: int | None = field(
        metadata=field_options(alias="sdpMLineIndex"), default=None, init=False
    )
    sdp_mid: str | None = field(
        metadata=field_options(alias="sdpMid"), default=None, init=False
    )
    user_fragment: str | None = field(
        metadata=field_options(alias="userFragment"), default=None, init=False
    )

    foundation: str | None = field(
        default=None, init=False, metadata=field_options(serialize="omit")
    )
    component: str | None = field(
        default=None, init=False, metadata=field_options(serialize="omit")
    )
    priority: int | None = field(
        default=None, init=False, metadata=field_options(serialize="omit")
    )
    address: str | None = field(
        default=None, init=False, metadata=field_options(serialize="omit")
    )
    protocol: str | None = field(
        default=None, init=False, metadata=field_options(serialize="omit")
    )
    port: int | None = field(
        default=None, init=False, metadata=field_options(serialize="omit")
    )
    type: str | None = field(
        default=None, init=False, metadata=field_options(serialize="omit")
    )
    tcp_type: str | None = field(
        metadata=field_options(alias="tcpType", serialize="omit"),
        default=None,
        init=False,
    )
    related_address: str | None = field(
        metadata=field_options(alias="relatedAddress", serialize="omit"),
        default=None,
        init=False,
    )
    related_port: str | None = field(
        metadata=field_options(alias="relatedPort", serialize="omit"),
        default=None,
        init=False,
    )

    relay_protocol: str | None = field(
        metadata=field_options(alias="relayProtocol", serialize="omit"),
        default=None,
        init=False,
    )
    url: str | None = field(
        default=None, init=False, metadata=field_options(serialize="omit")
    )

    @classmethod
    def __pre_deserialize__(cls, d: dict[Any, Any]) -> dict[Any, Any]:
        """Per spec candidate serializes and deserialises to/from RTCIceCAndidateInit.

        To invoke the toJSON() operation of the RTCIceCandidate interface,
        run the following steps:
        1. Let json be a new RTCIceCandidateInit dictionary.
        2. For each attribute identifier attr in candidate, sdpMid,
            sdpMLineIndex, usernameFragment:
        3. Let value be the result of getting the underlying value of the attribute
            identified by attr, given this RTCIceCandidate object.
        4. Set json[attr] to value.
        5. Return json.

        This method reverses that logic.
        """
        candidate_init = RTCIceCandidateInit.from_dict(d)
        return {"rtc_ice_candidate_init": candidate_init.to_dict()}

    def __post_init__(self) -> None:
        """Initialize the class.

        Spec compliance: If both the sdpMid and sdpMLineIndex members of
        candidateInitDict are null, throw a TypeError.
        """
        ric_init = self.rtc_ice_candidate_init
        if (
            ric_init
            and ric_init.candidate != ""
            and ric_init.sdp_m_line_index is None
            and ric_init.sdp_mid is None
        ):
            msg = "sdp_m_line_index and sdp_mid cannot both be null."
            raise TypeError(msg)

        if ric_init:
            # Attributes are readonly so set them via __setattr__
            object.__setattr__(self, "candidate", ric_init.candidate)
            object.__setattr__(self, "sdp_mid", ric_init.sdp_mid)
            object.__setattr__(self, "sdp_m_line_index", ric_init.sdp_m_line_index)
            object.__setattr__(self, "user_fragment", ric_init.user_fragment)
        else:
            object.__setattr__(self, "candidate", "")
