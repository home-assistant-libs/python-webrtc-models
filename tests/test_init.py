"""WebRTC models tests."""

from __future__ import annotations

from typing import TYPE_CHECKING

import orjson
import pytest

from tests import load_fixture
from webrtc_models import RTCConfiguration, RTCIceCandidate, RTCIceServer

if TYPE_CHECKING:
    from mashumaro.mixins.orjson import DataClassORJSONMixin
    from syrupy import SnapshotAssertion


@pytest.mark.parametrize(
    ("clazz", "filename"),
    [
        # RTCIceServer
        (RTCIceServer, "RTCIceServer_only_urls_string.json"),
        (RTCIceServer, "RTCIceServer_only_urls_list.json"),
        (RTCIceServer, "RTCIceServer_urls_string.json"),
        (RTCIceServer, "RTCIceServer_urls_list.json"),
        # RTCConfiguration
        (RTCConfiguration, "RTCConfiguration_empty.json"),
        (RTCConfiguration, "RTCConfiguration_one_iceServer.json"),
        (RTCConfiguration, "RTCConfiguration_multiple_iceServers.json"),
        # RTCIceCandidate
        (RTCIceCandidate, "RTCIceCandidate_end.json"),
        (RTCIceCandidate, "RTCIceCandidate_candidate.json"),
    ],
)
def test_decoding_and_encoding(
    snapshot: SnapshotAssertion,
    clazz: type[DataClassORJSONMixin],
    filename: str,
) -> None:
    """Test decoding/encoding."""
    file_content = load_fixture(filename)
    instance = clazz.from_json(file_content)
    assert instance == snapshot(name="dataclass")

    file_content_dict = orjson.loads(file_content)
    instance_dict = instance.to_dict()
    assert instance_dict == snapshot(name="dict")

    # Verify json
    assert instance.to_json() == orjson.dumps(file_content_dict).decode()

    # Verify dict
    assert instance_dict == file_content_dict
    assert instance == clazz.from_dict(instance_dict)


def test_no_mid_and_mlineindex() -> None:
    """Test spd_mid and sdp_multilineindex raises TypeError."""
    file_content = load_fixture("RTCIceCandidate_invalid.json")
    with pytest.raises(TypeError):
        RTCIceCandidate.from_json(file_content)


def test_empty_candidate_creation() -> None:
    """Test empty candidate creation."""
    file_content = load_fixture("RTCIceCandidate_end.json")
    file_content_dict = orjson.loads(file_content)
    rtc_ice_end = RTCIceCandidate()
    assert rtc_ice_end.to_dict() == file_content_dict
