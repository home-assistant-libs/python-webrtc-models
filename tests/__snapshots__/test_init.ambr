# serializer version: 1
# name: test_decoding_and_encoding[RTCConfiguration-RTCConfiguration_empty.json][dataclass]
  dict({
    'ice_servers': list([
    ]),
  })
# ---
# name: test_decoding_and_encoding[RTCConfiguration-RTCConfiguration_empty.json][dict]
  dict({
  })
# ---
# name: test_decoding_and_encoding[RTCConfiguration-RTCConfiguration_multiple_iceServers.json][dataclass]
  dict({
    'ice_servers': list([
      dict({
        'credential': None,
        'urls': 'stun:stun.home-assisant.io:80',
        'username': None,
      }),
      dict({
        'credential': 'credential',
        'urls': list([
          'stun:stun.home-assisant.io:80',
          'stun:stun.l.google.com:19302',
        ]),
        'username': 'username',
      }),
    ]),
  })
# ---
# name: test_decoding_and_encoding[RTCConfiguration-RTCConfiguration_multiple_iceServers.json][dict]
  dict({
    'iceServers': list([
      dict({
        'urls': 'stun:stun.home-assisant.io:80',
      }),
      dict({
        'credential': 'credential',
        'urls': list([
          'stun:stun.home-assisant.io:80',
          'stun:stun.l.google.com:19302',
        ]),
        'username': 'username',
      }),
    ]),
  })
# ---
# name: test_decoding_and_encoding[RTCConfiguration-RTCConfiguration_one_iceServer.json][dataclass]
  dict({
    'ice_servers': list([
      dict({
        'credential': None,
        'urls': 'stun:stun.home-assisant.io:80',
        'username': None,
      }),
    ]),
  })
# ---
# name: test_decoding_and_encoding[RTCConfiguration-RTCConfiguration_one_iceServer.json][dict]
  dict({
    'iceServers': list([
      dict({
        'urls': 'stun:stun.home-assisant.io:80',
      }),
    ]),
  })
# ---
# name: test_decoding_and_encoding[RTCIceCandidate-RTCIceCandidate_candidate.json][dataclass]
  dict({
    'address': None,
    'candidate': '3932168448 1 udp 1694498815 1.2.3.4 10676 typ srflx raddr 0.0.0.0 rport 37566',
    'component': None,
    'foundation': None,
    'port': None,
    'priority': None,
    'protocol': None,
    'related_address': None,
    'related_port': None,
    'relay_protocol': None,
    'rtc_ice_candidate_init': dict({
      'candidate': '3932168448 1 udp 1694498815 1.2.3.4 10676 typ srflx raddr 0.0.0.0 rport 37566',
      'sdp_m_line_index': 0,
      'sdp_mid': None,
      'user_fragment': None,
    }),
    'sdp_m_line_index': 0,
    'sdp_mid': None,
    'tcp_type': None,
    'type': None,
    'url': None,
    'user_fragment': None,
  })
# ---
# name: test_decoding_and_encoding[RTCIceCandidate-RTCIceCandidate_candidate.json][dict]
  dict({
    'candidate': '3932168448 1 udp 1694498815 1.2.3.4 10676 typ srflx raddr 0.0.0.0 rport 37566',
    'sdpMLineIndex': 0,
    'sdpMid': None,
    'userFragment': None,
  })
# ---
# name: test_decoding_and_encoding[RTCIceCandidate-RTCIceCandidate_end.json][dataclass]
  dict({
    'address': None,
    'candidate': '',
    'component': None,
    'foundation': None,
    'port': None,
    'priority': None,
    'protocol': None,
    'related_address': None,
    'related_port': None,
    'relay_protocol': None,
    'rtc_ice_candidate_init': dict({
      'candidate': '',
      'sdp_m_line_index': None,
      'sdp_mid': None,
      'user_fragment': None,
    }),
    'sdp_m_line_index': None,
    'sdp_mid': None,
    'tcp_type': None,
    'type': None,
    'url': None,
    'user_fragment': None,
  })
# ---
# name: test_decoding_and_encoding[RTCIceCandidate-RTCIceCandidate_end.json][dict]
  dict({
    'candidate': '',
    'sdpMLineIndex': None,
    'sdpMid': None,
    'userFragment': None,
  })
# ---
# name: test_decoding_and_encoding[RTCIceServer-RTCIceServer_only_urls_list.json][dataclass]
  dict({
    'credential': None,
    'urls': list([
      'stun:stun.home-assisant.io:80',
      'stun:stun.l.google.com:19302',
    ]),
    'username': None,
  })
# ---
# name: test_decoding_and_encoding[RTCIceServer-RTCIceServer_only_urls_list.json][dict]
  dict({
    'urls': list([
      'stun:stun.home-assisant.io:80',
      'stun:stun.l.google.com:19302',
    ]),
  })
# ---
# name: test_decoding_and_encoding[RTCIceServer-RTCIceServer_only_urls_string.json][dataclass]
  dict({
    'credential': None,
    'urls': 'stun:stun.home-assisant.io:80',
    'username': None,
  })
# ---
# name: test_decoding_and_encoding[RTCIceServer-RTCIceServer_only_urls_string.json][dict]
  dict({
    'urls': 'stun:stun.home-assisant.io:80',
  })
# ---
# name: test_decoding_and_encoding[RTCIceServer-RTCIceServer_urls_list.json][dataclass]
  dict({
    'credential': 'credential',
    'urls': list([
      'stun:stun.home-assisant.io:80',
      'stun:stun.l.google.com:19302',
    ]),
    'username': 'username',
  })
# ---
# name: test_decoding_and_encoding[RTCIceServer-RTCIceServer_urls_list.json][dict]
  dict({
    'credential': 'credential',
    'urls': list([
      'stun:stun.home-assisant.io:80',
      'stun:stun.l.google.com:19302',
    ]),
    'username': 'username',
  })
# ---
# name: test_decoding_and_encoding[RTCIceServer-RTCIceServer_urls_string.json][dataclass]
  dict({
    'credential': 'credential',
    'urls': 'stun:stun.home-assisant.io:80',
    'username': 'username',
  })
# ---
# name: test_decoding_and_encoding[RTCIceServer-RTCIceServer_urls_string.json][dict]
  dict({
    'credential': 'credential',
    'urls': 'stun:stun.home-assisant.io:80',
    'username': 'username',
  })
# ---
